blog/authenticating_a_rails_api_part_1.md

```diff
@@ -0,0 +1,295 @@
+  * If a user with that auth token exists and the auth token is not expired, the request proceeds.
+  * Otherwise, the server returns a `401 - Unauthorized` response.
+3. User regenerates a token
+  * If their token has expired, the user submits their email and password to a `POST /authentications` endpoint.
+  * A new token is generated for that user and is returned for use in subsequent requests.
+
+We'll tackle parts 1 and 3 today.  Follow along on [this repo](https://github.com/IntrepidPursuits/api-authentication-demo).
+
+### Tools we'll use
+
+- [ActiveModel::Serializers v. 0.8.3](https://github.com/rails-api/active_model_serializers/tree/0-8-stable)
+- [Monban](https://github.com/halogenandtoast/monban) and [Warden](https://github.com/hassox/warden)
+- RSpec and [json_spec](https://github.com/collectiveidea/json_spec) gem for testing
+- [Versionist](https://github.com/bploetz/versionist) for versioning the API
+
+### Let's get started
+
+#### Step 1: Write a test
+
+Because we're good little TDD-ers, we'll start with an integration test for our "happy path" case -- that is, the case where everything goes right:
+
+```ruby
+# /spec/requests/v1/users_requests_spec.rb
+
+require 'rails_helper'
+
+describe 'Users endpoints' do
+  describe 'POST /users' do
+    context 'with valid email and password' do
+      it 'returns JSON for a user' do
+        user_attributes = {
+          user: {
+            email: user.email,
+            password: user.password_digest
+          }
+        }.to_json
+
+        post(users_url, user_attributes, accept_headers)
+
+        expect(response).to have_http_status :ok
+        expect_response_to_include_user_info
+      end
+    end
+  end
+
+  def user
+    @user ||= build(:user)
+  end
+
+  def expect_response_to_include_user_info
+    expect(json_value_at_path('user/email')).to eq user.email
+    expect(json_value_at_path('user/authentication_token')).to be
+    expect(json_value_at_path('user/authentication_token_expires_at')).to be
+  end
+end
+```
+
+We're making use of a couple of helper methods in our test that we can define in a `Helpers::Requests` module, namely:
+
+* `accept_headers` to generate a header with our vendor prefix and API version number
+* `json_value_at_path` for validating that JSON responses contain the correct values
+
+These rely on the `json_spec` gem's [helper methods](https://github.com/collectiveidea/json_spec#rspec), namely `parse_json`.
+
+```ruby
+# spec/support/helpers/requests.rb
+
+module Helpers
+  module Requests
+    def body
+      response.body
+    end
+
+    def json_value_at_path(path = '')
+      parse_json(body, path)
+    end
+
+    def api_version
+      1
+    end
+
+    def accept_header
+      "application/vnd.authentication-demo-app.com; version=#{api_version}"
+    end
+
+    def accept_headers
+      { 'Accept' => accept_header,
+        'Content-Type' => 'application/json' }
+    end
+  end
+end
+```
```

Comment:
I realize we're targeting beginner to intermediate rails users, but no guarantee they know rspec - some discussion about what you're doing in these tests wouldn't hurt.